meliuz/meliorator
# chore(bump): Cleo from 0.6 to 2.1

# Description
The dependabot is requesting to update the cleo library to a up to date version. To do this, we did need to refact all commands code to attain to the new methods and parameters. We upgrade the following dependecies:
- minimum python to 3.9
- cleo: 0.6 -> 2.1

We decided to go for 2.1 because was the newest one, besides the lack of documentation. The main changes was in the self.call parameters, where the `options` changed to args, which is a string parameter now, and the command attributes that now are class attributes instead of comments.  For references i used the poetry library ([example](https://github.com/python-poetry/poetry/blob/master/src/poetry/console/commands/add.py)) . The format `namespace:command` was chagend to `namespace command` but the old versions are still available as a alieases.

There is also a bug in the self.call when the command receives an argument as first parameter. To workaround this we need to pass the command name also in the args parameter. To more details see: https://github.com/python-poetry/cleo/issues/130

Some updates only in dev environment:
- flexmock: 0.9 to 0.11
- pytest: 3.5 -> 7.4
- pytest-mock: 1.6 to 3.12

⚠️ For now the testing environment is broken because we need to adapt the environment to run locally, maybe with docker or something. And remove deprecated environments, like Python <3.9. We need to address this issue later. For now, i only update the flexmock and pytest to run the tests in the command module.

# Setup
- Install [poetry](https://python-poetry.org/docs/#installing-with-pipx)
- Install [psql](https://www.timescale.com/blog/how-to-install-psql-on-mac-ubuntu-debian-windows/)
- Create a new virtual environment with `poetry env use $(which python3)` 
- If it didnt not enter in the venv, you can do with `poetry shell`
- Install psycopg2 with `pip install psycopg2-binary`
- Execute `poetry build` to generate the package wheel
- Execute `pip install  dist/meliorator-0.9.24-py3-none-any.whl` to install the orator inside the virtual env.
- Run `orator list` to ensure that everything is ok
- Create a file `database/orator.py` with the following content:
```python
from orator import DatabaseManager


configuration = {
    'main': {
        'driver': 'postgres',
        'host': 'localhost',
        'database': 'orator_test',
        'user': 'postgres',
        'password': 'postgres',
        'prefix': '',
        'use_qmark': True,
    },
}

db = DatabaseManager(configuration)

DATABASES = configuration
```

- Considering the postgres is running in localhost and port 5432, create a testing database with:
```sh
PGPASSWORD=postgres psql -h localhost -U postgres -c 'CREATE DATABASE orator_test'
```

- You can drop the database with
```
PGPASSWORD=postgres psql -h localhost -U postgres -c 'DROP DATABASE orator_test'

```

# Tests
## Creating migration files (this will test migrate:install too)
- Run `orator make:migration test`
- A file in a `migrations` folder should have been created with no content in `up` or `down` methods.
- Run `orator make:migration test -t test_table`
- A file should have beem created with self.schema.table executions
- Run `orator make:migration test -t test_table2 --create`
- A file shoud have been created with self.schema.create executions

## Running migrations
- Run `orator migrate -c database/orator.py -f`
- Open postgres with `PGPASSWORD=postgres psql -h localhost -U postgres -d orator_test`
- Run `\d`
- The tables `migrations` and `test_table2` should have been created
- Run `quit` to quit the psql shell

## Running rollbacks (almost same effect as migrate:reset
- Run `orator migrate:rollback -c database/orator.py -f`
- Enter again in the database. All tables but `migrations` must been removed

## Recreating the database
- Run `orator migrate:refresh -c database/orator.py`
- Enter again in the database. All tables must be recreated.

## Running seeds
- Create a seed file with `orator make:seed test`
- Two files should have been created in `seeds` folder: `database_seeder.py` and `test.py`
- Modify the test.py file with the following:
```python
from orator.seeds import Seeder


class Test(Seeder):

    def run(self):
        """
        Run the database seeds.
        """
        self.db.table('test_table2').insert({
            'id': 1,
        })
```
- Modify the database_seeder.py with the following:
```python
from orator.seeds import Seeder
from .test import Test

class DatabaseSeeder(Seeder):

    def run(self):
        """
        Run the database seeds.
        """
        self.call(Test)
```
- Run `orator db:seed -c database/orator.py -f`
- Enter in the database and run `SELECT * FROM test_table2;`
- It should return the inserted object

## Recreates everything at once
- Run `orator migrate:refresh --seed -c database/orator.py`
- It should run rollbacks, migrations and seeds successfully.
